---
title: "Destin Buenrostro Mouse"
output:
  rmarkdown::html_document:
    highlight: pygments
    toc: true
---

For this example we begin at this step 3.1. Load Destin R package.  Previous steps are described for reference.

# 1. Data
- Mouse scATAC-seq from Buenrostro et al. 
- GEO GSE65360  
- 2 cell types
- Fluidigm microfluidics platform
- 192 cells (20 in this toy example)
- original fastq files correspond to individual cells

# 2. Bioinformatics pipeline

## 2.1. Set up parameters, paths, and directories

### 2.1.1. designate inputs and parameters
```{bash eval=FALSE}
outDir=~/scATACseq/pipelinePractice/BuenrostroMouse
SRRFile=$outDir/SRR_Acc_ListBuenrostroMouse.txt
sampleName=BuenrostroMouse
srcDir=~/scATACseq/scATACseqPipeline/src
model=mm10
cellData=$outDir/SraRunTableBuenrostroMouse.txt
```


### 2.1.2. create directories
```{bash eval=FALSE}
mkdir -p $outDir/temp
mkdir -p $outDir/bam
mkdir -p $outDir/peaks
mkdir -p $outDir/fastq
```


### 2.1.3. set model specific parameters for  mm10 
these should be set specific to your system
```{bash eval=FALSE} 
export genomeBowtie=/proj/seq/data/MM10_UCSC/Sequence/Bowtie2Index/genome
export genomeMacs2=mm;
export blacklistFile=~/scATACseq/scATACseqPipeline/data/blacklist/ENCFF547MET.bed.gz
export genomeBedtools=/proj/seq/data/MM10_UCSC/Sequence/Chromosomes
```


### 2.1.4. load modules
below is an example, but this will need to be customized so that calls to dependencies are included in your $PATH
```{bash eval=FALSE}
module add sratoolkit
module add cutadapt
module add bowtie2
module add samtools
export picardPath=~/applications/picard/build/libs
module add macs/2016-02-15
module add bedtools
```


## 2.2. Download fastq files for toy example
```{bash eval=FALSE}
SRRNames=( $( cat $SRRFile ) )
fastq-dump "${SRRNames[@]}" --split-files -O $outDir/fastq --gzip
``` 


## 2.3. Align reads
- cut adapters 
- align 
- sam to bam
- sort
- Add read group and index
- mark duplicates
- remove mitochondrial, unmapped and chr Y
- adjust for Tn5 insertion
- alignment quality >= 30
- index

```{bash eval=FALSE}
for cellID in "${cellIDs[@]}"
do
  $srcDir/fastqToBam.sh \
    -c $cellID  \
    -s $sampleName \
    -o $outDir \
    -g $genomeBowtie
done
```


## 2.4. Call Peaks
- call peaks (p < 0.01)
- filter blacklist

```{bash eval=FALSE}
$srcDir/callPeaks.sh \
  -s $sampleName \
  -o $outDir \
  -g $genomeBedtools \
  -l $blacklistFile \
  -q $genomeMacs2
```

# 3. Destin

## 3.1. Load Destin R package
```{r message=FALSE, warning=FALSE}
#install.packages("~/Documents/gitRepos/destin/package", repos = NULL, type = "source")
library(destin, quietly = T)
```


## 3.2. set parameters
```{r}
sampleName = "BuenrostroMouse"
outDir = "~/Documents/gitRepos/destin/practice/BuenrostroMouse"
cellDataFile = file.path(outDir, "SraRunTableBuenrostroMouse.txt")
model = "mm10"
```


## 3.3. create RSE from bam and peaks
```{r}
bamDir = file.path(outDir, "bam")
peaksDir = file.path(outDir, "peaks")
bedFile = file.path(peaksDir, 
    paste0(sampleName, "_peaks.blacklist_removed.narrowPeak"))

extraCols_narrowPeak <- c(signalValue = "numeric", pValue = "numeric",
                          qValue = "numeric", peak = "integer")
bedData = import(bedFile, format = "BED",
               extraCols = extraCols_narrowPeak)

bamFiles = dir( bamDir, pattern = "final.bam" )
bamFiles = bamFiles[!grepl("bai", bamFiles)]

rse = createRSE(bamDir, bamFiles, bedData)
```


## 3.4. append experimental information if available
in this example we append the SRA Run table from SRA
```{r}
cellData = fread(cellDataFile) 
colMerged = merge(rse@colData, cellData, by.x = "cellID", by.y = "Run")
colData(rse) = colMerged 
```

## 3.5. annotate regions
Annotate the ranges with TSS distance and DHS frequency 
```{r}
rse = annotateRSE(rse, model, DHSfile)
```

## 3.6. QC
Destin begins with quality control, retaining chromatin regions accessible in at least 5 cells, and retaining cells with total number of chromatin accessible regions within 3 standard deviations from median (robustly calculated using median absolute deviations).
```{r}
rse = doQC(rse)
```

## 3.7. Determine number of clusters
We calculate the number of clustes using the model-based systematic elbow.  Here we find 2 clusters.
```{r}
set.seed(10)
clusterEst = estimateNClusters(rse,   clusterVec = 1:5)
qplot(data = clusterEst$logLikes, x = nClusters, y = logLike) 
```

## 3.8. Destin
We run Destin with default settings aside from lowering the number of principle components (since we only have 17 cells post-QC).  Optimal tuning parameters for the toy example are displayed in the summary: TSS weights (1, 2), higher weight to distal regulatory elements; DHS weights (1, 1), unweighted DHS frequency; and 3 principle components.
```{r warning=FALSE}
nClusters = clusterEst$nClustersEstimate
PCrange = 3:10
TSSWeightsList = list(c(1,2), c(1,1.5), c(1,1.25), c(1,1))
DHSWeightsList = list(c(1,1), c(1,2), c(1,3), c(1,5))
results = destinGrid (rse, sampleName, 
            PCrange = PCrange,
            TSSWeightsList = TSSWeightsList,
            DHSWeightsList = DHSWeightsList,
            nClusters = nClusters)
results
```

Since Buenrostro et al. provided cell labels, we can compute cluster purity
```{r}
clusterTable = table(results$cluster$cluster, rse$cell_type)
purity = sum( apply(clusterTable, 1, max) ) / sum(clusterTable)
clusterTable
paste0("purity = ", purity)
```


## 3.9. Differential Accessibility
Calculate differential accessibility
```{r}
rse = getDiffAccess(rse, clusterMap = results$cluster)
```

Results of differential accessibility (top 10 p-values per cluster)
```{r}
rowRanges(rse)[order(rowRanges(rse)$pValsCor_1, decreasing = F)][1:10, c("feature", "pValsCor_1", "log2FC_1")] 

rowRanges(rse)[order(rowRanges(rse)$pValsCor_2, decreasing = F)][1:10, c("feature", "pValsCor_2", "log2FC_2")] 
```
